File: api/src/spp_extras_api/utils/quests.py
import json
from from_root import from_root
with open(from_root('data/questRaceZeros.json'), 'r') as json_file:
    quest_race_zeros = json.load(json_file)

# Organize completed quests by character
def format_completed_quests(regular, daily, weekly, monthly):
    all = {}
    def add_quest(quest, type):
        guid = str(quest['guid'])
        questId = str(quest['quest'])
        if str(guid) not in all:
            all[str(guid)] = {
                'regular': {},
                'daily': {},
                'weekly': {},
                'monthly': {}
            }

        if type == 'regular':
            all[guid]['regular'][questId] = quest
        elif type == 'daily' and daily:
            all[guid]['daily'][questId] = quest
        elif type == 'weekly':
            all[guid]['weekly'][questId] = quest
        elif type == 'monthly' and monthly:
            all[guid]['monthly'][questId] = quest

    for q in regular:
        add_quest(q, 'regular')
    for q in weekly:
        add_quest(q, 'weekly')
    if daily:
        for q in daily:
            add_quest(q, 'daily')
    if monthly:
        for q in monthly:
            add_quest(q, 'monthly')

    return all


# Organize template quests by faction
def format_template_quests(quests):
    all = {
        'alliance': {},
        'horde': {},
        'both': {}
    }

    alliance = [1, 4, 5, 8, 64, 65, 68, 77, 1101, 1024]
    horde = [2, 16, 32, 128, 130, 144, 162, 176, 178, 512, 514, 690]
    both = [0, 255, 1791]

    for quest in quests:
        required_races = quest['requiredraces']
        entry = str(quest['entry'])

        # Some quests with requiredraces of 0 are faction specific
        if entry in quest_race_zeros:
            quest_faction = quest_race_zeros[entry]
            all[quest_faction][entry] = quest
            continue

        # Add other quests based on requiredraces
        if required_races in alliance:
            all['alliance'][entry] = quest
        elif required_races in horde:
            all['horde'][entry] = quest
        elif required_races in both:
            all['both'][entry] = quest

    return all

File: src/utils/quests.ts
import _questRaceZeros from '../../data/quests/questRaceZeros.json';
import {
  AllCharacters,
  AllCompletedQuests,
  AllTemplateQuests,
  CompletedQuests,
  RawComplRepeatQuests,
  RawComplRegQuests,
  Quest,
  QuestRaceZeros,
  QuestType,
  RawTemplateQuests,
} from "../types";

const questRaceZeros = _questRaceZeros as QuestRaceZeros;

// Organize completed quests by character
export const formatCompletedQuests = (
  regular: RawComplRegQuests,
  daily: RawComplRepeatQuests,
  weekly: RawComplRepeatQuests,
  monthly: RawComplRepeatQuests,
) => {
  const all: CompletedQuests = {};
  const addQuest = (quest: Quest, type: QuestType) => {
    const guid = quest.guid.toString();
    const questID = quest.quest.toString();
    if (!all[guid]) {
      all[guid] = {
        regular: {},
        daily: {},
        weekly: {},
        monthly: {}
      };
    }

    if (type === 'regular') all[guid]['regular'][questID] = quest;
    else if (type === 'daily' && daily) all[guid]['daily'][questID] = quest;
    else if (type === 'weekly') all[guid]['weekly'][questID] = quest;
    else if (type === 'monthly' && monthly) all[guid]['monthly'][questID] = quest;
  }

  regular.forEach(q => addQuest(q, 'regular'));
  if (daily) daily.forEach(q => addQuest(q, 'daily'));
  weekly.forEach(q => addQuest(q, 'weekly'));
  if (monthly) monthly.forEach(q => addQuest(q, 'monthly'));

  return all;
}

// Organize completed quests by faction
export const formatComplFactionQuests = (
  characters: AllCharacters,
  completedQuests: CompletedQuests,
) => {
  const all: AllCompletedQuests = { alliance: {}, horde: {} };

  // Gather quests completed only by player characters
  Object.values(characters.alliance).forEach(c => {
    all['alliance'][c.guid] = completedQuests[c.guid];
  });

  Object.values(characters.horde).forEach(c => {
    all['horde'][c.guid] = completedQuests[c.guid];
  });

  return all;
}

// Organize template quests by faction
export const formatTemplateQuests = (quests: RawTemplateQuests) => {
  const all: AllTemplateQuests = {
    alliance: {},
    horde: {},
    neutral: {}
  };

  const alliance = [1, 4, 5, 8, 64, 65, 68, 77, 1101, 1024];
  const horde = [2, 16, 32, 128, 130, 144, 162, 176, 178, 512, 514, 690];
  const neutral = [0, 255, 1791];
  quests.forEach(quest => {
    const reqRaces = quest.RequiredRaces;
    const entry = quest.entry;

    // Some quests with RequiredRaces of 0 are faction specific
    const questFaction = questRaceZeros[entry];
    if (questFaction) {
      all[questFaction][entry] = quest;
      return;
    }

    // Add other quests based on RequiredRaces
    if (alliance.includes(reqRaces)) all['alliance'][entry] = quest;
    else if (horde.includes(reqRaces)) all['horde'][entry] = quest;
    else if (neutral.includes(reqRaces)) all['neutral'][entry] = quest;
  });

  return all;
}
